vortex-layout/src/layouts/dict/writer/mod.rs

```diff
@@ -0,0 +1,188 @@
+    fn push_chunk(
+        &mut self,
+        segment_writer: &mut dyn crate::segments::SegmentWriter,
+        chunk: ArrayRef,
+    ) -> VortexResult<()> {
+        match self {
+            Self::Uninit(args) => {
+                let compressed = BtrBlocksCompressor.compress(&chunk)?;
+                *self = if !compressed.is_encoding(DictEncoding.id()) {
+                    Self::Writer(args.child_strategy.new_writer(&args.ctx, &args.dtype)?)
+                } else if args.options.repeat {
+                    Self::Writer(
+                        repeating::RepeatingDictLayoutWriter::new(
+                            args.ctx.clone(),
+                            &args.dtype,
+                            args.child_strategy.clone(),
+                            args.values_strategy.clone(),
+                            args.options.constraints.clone(),
+                        )
+                        .boxed(),
+                    )
+                } else {
+                    Self::Writer(
+                        single::DictLayoutWriter::new(
+                            args.ctx.clone(),
+                            &args.dtype,
+                            args.child_strategy.clone(),
+                            args.values_strategy.clone(),
+                            args.options.constraints.clone(),
+                        )
+                        .boxed(),
+                    )
+                };
+                self.push_chunk(segment_writer, chunk)
+            }
+            Self::Writer(writer) => writer.push_chunk(segment_writer, chunk),
+        }
+    }
+
+    fn flush(
+        &mut self,
+        segment_writer: &mut dyn crate::segments::SegmentWriter,
+    ) -> VortexResult<()> {
+        match self {
+            Self::Uninit(_) => vortex_bail!("flush called before push_chunk"),
+            Self::Writer(writer) => writer.flush(segment_writer),
+        }
+    }
+
+    fn finish(
+        &mut self,
+        segment_writer: &mut dyn crate::segments::SegmentWriter,
+    ) -> VortexResult<Layout> {
+        match self {
+            Self::Uninit(_) => vortex_bail!("finish called before push_chunk"),
+            Self::Writer(writer) => writer.finish(segment_writer),
+        }
+    }
+}
+
+#[derive(Debug, Clone, rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)]
+pub struct DictLayoutMetadata {
+    pub codes_ptype: PType,
+}
+
+fn dict_layout(values: Layout, codes: Layout) -> VortexResult<Layout> {
+    let codes_ptype = codes.dtype().try_into()?;
+    let metadata = Bytes::copy_from_slice(
+        &RkyvMetadata(DictLayoutMetadata { codes_ptype })
+            .serialize()
+            .ok_or_else(|| vortex_err!("could not serialize dict layout metadata"))?,
+    );
+    Ok(Layout::new_owned(
+        "dict".into(),
+        LayoutVTableRef::new_ref(&DictLayout),
+        values.dtype().clone(),
+        codes.row_count(),
+        vec![],
+        vec![values, codes],
+        Some(metadata),
```

Comment:
Might not be in your PR, but metadata should be a `Vec<u8>` not `ByteBuffer` since we provide no alignment guarantees